src/osd/pg_scrubber.cc

```diff
@@ -0,0 +1,2139 @@
+  m_epoch_start = m_pg->info.history.same_interval_since;
+  m_replica_request_priority = msg->high_priority ? Scrub::scrub_prio_t::high_priority
+						  : Scrub::scrub_prio_t::low_priority;
+  m_flags.priority = msg->priority ? msg->priority : m_pg->get_scrub_priority();
+
+  preemption_data.reset();
+  preemption_data.force_preemptability(msg->allow_preemption);
+
+  replica_scrubmap_pos.reset();
+
+  // make sure the FSM is at NotActive
+  m_fsm->assert_not_active();
+
+  m_osds->queue_for_rep_scrub(m_pg, m_replica_request_priority, m_flags.priority);
+}
+
+void PgScrubber::replica_scrub(epoch_t epoch_queued)
+{
+  dout(10) << __func__ << ": " << m_pg->pg_id << " epoch queued: " << epoch_queued
+	   << dendl;
+  dout(20) << __func__ << " m_epoch_start: " << m_epoch_start
+	   << " better be >= " << m_pg->info.history.same_interval_since << dendl;
+  dout(20) << __func__ << " m_is_deep: " << m_is_deep << dendl;
+
+  if (m_pg->pg_has_reset_since(epoch_queued)) {
+    dout(10) << "replica_scrub(epoch,) - reset!" << dendl;
+    send_epoch_changed();
+    return;
+  }
+
+  if (was_epoch_changed()) {
+    dout(10) << "replica_scrub(epoch,) - epoch!" << dendl;
+    send_epoch_changed();
+    return;
+  }
+  ceph_assert(!is_primary());  // as should have been caught by the epoch-changed check
+
+  send_start_replica();
+}
+
+void PgScrubber::replica_scrub_resched(epoch_t epoch_queued)
+{
+  dout(10) << __func__ << ": " << m_pg->pg_id << " epoch queued: " << epoch_queued
+	   << dendl;
+
+  if (m_pg->pg_has_reset_since(epoch_queued)) {
+    dout(10) << "replica_scrub(epoch,) - reset!" << dendl;
+    send_epoch_changed();
+    return;
+  }
+
+  if (was_epoch_changed()) {
+    dout(10) << __func__ << " epoch changed!" << dendl;
+    send_epoch_changed();
+    return;
+  }
+  ceph_assert(!is_primary());  // as should have been caught by the epoch-changed check
+
+  send_sched_replica();
+}
+
+void PgScrubber::set_op_parameters(requested_scrub_t& request)
+{
+  dout(10) << __func__ << " input: " << request << dendl;
+
+  m_flags.check_repair = request.check_repair;
+  m_flags.auto_repair = request.auto_repair || request.need_auto;
+  m_flags.required = request.req_scrub || request.must_scrub;
+
+  m_flags.priority = (request.must_scrub || request.need_auto)
+		       ? get_pg_cct()->_conf->osd_requested_scrub_priority
+		       : m_pg->get_scrub_priority();
+
+  state_set(PG_STATE_SCRUBBING);
+
+  // will we be deep-scrubbing?
+  if (request.must_deep_scrub || request.need_auto || request.time_for_deep) {
+    state_set(PG_STATE_DEEP_SCRUB);
+  }
+
+  if (request.must_repair || m_flags.auto_repair) {
+    state_set(PG_STATE_REPAIR);
+  }
+
+  // the publishing here seems to be required for tests synchronization
+  m_pg->publish_stats_to_osd();
+  m_flags.deep_scrub_on_error = request.deep_scrub_on_error;
+  request = requested_scrub_t{};
+}
+
+/**
+ *  RRR \todo ask why we collect from acting+recovery+backfill, but use the size of
```

Comment:
Are you asking about line 1166 below?  Most likely, this was a mistake from when acting_recovery_backfill was added -- you should probably swap it over.